### DISCUSSION TITLE
[Question] Program correctness for recursive functions

### DISCUSSION THREAD

**Discussion by pingpingy1** [prosyslab-classroom/cs424-program-reasoning#319] [Q&A]
Is it possible to use the techniques we have used to prove the correctness of recursive functions?
For example, the following function is recursive:
```
@pre: n >= 0
@post: isNthFibonacciNumber(ret, n)
let rec fibo = function
    | 0 -> 0
    | 1 -> 1
    | n -> (fibo (n - 1)) + (fibo (n - 2))
```
To prove the correctness of this function, my instinct is to use induction on `n`, which might work for this particular case since it is a simple function with only one integer input.

However, for more complex functions like merge sort:
```
@pre: 0 <= l, r < length(a)
@post: forall i, j. l <= i <= j < r -> a[i] <= a[j]
let rec mergeSort a l r =
    if l >= r then return;
    let mid = (l + r) / 2 in
    (Something something)
    let _ = mergeSort a l mid in
    let _ = mergeSort a mid r in
    (Something something)
```
such simple mathematical induction seems inapplicable.
It feels like we need to know how `mergeSort` functions in order to reason about its function, let alone its correctness.
I attempted to say that if `mergeSort` correctly sorts the left half and it correctly sorts the right half, then it satisfies the postcondition.
However, this argument begs the question, i.e., it assumes the postcondition that I am trying to prove, making it vacuous.

In short, how do we prove the partial correctness of recursive functions?

Thanks.

**Comment by KihongHeo** [comment]
Hi Heewon. That is a great question.

In principle, loop is conceptually the same as recursion. So you can transform any algorithm with loops to one with recursion.

In such cases, loop invariants correspond to the function's pre/post conditions. In your example, 
```
@pre: 0 <= l, r < length(a)
@post: forall i, j. l <= i <= j < r -> a[i] <= a[j]
```
is such an invariant. So it is used as an induction hypothesis at the recursive calls. Since you assume the above pre, and post are correct, you can complete the proof by assuming the following as you mentioned:
> if `mergeSort` correctly sorts the left half and it correctly sorts the right half, then it satisfies the postcondition.

Notice that this is not a vacuous conclusion. This is the beauty of recursion: program structures (i.e., recursion) match with mathematical proofs (i.e., induction). That is why it is relatively easier to check the functional correctness of programs written in higher-level functional languages such as OCaml, Haskell, etc.

Then, what happens if a wrong pair of pre and post is given? Does that lead to an incorrect conclusion? No. Since the proof rules we discussed in the lecture are sound, if your pre and post are wrong, e.g.,
```
@pre: 0 <= l, r < length(a)
@post: forall i, j. l <= i <= j < r -> a[i] > a[j]     //  > should be <=
```
then you will not complete your proof using Hoare logic.

**Comment by pingpingy1** [comment]
Oh, okay.
So to summarize, the following procedure would be enough to prove the partial correctness of the `mergeSort` function, right?
1. Given `@pre`, if the function returns immediately, it satisfies `@post`
2. Given `@pre`, if the function does not return immediately, it enters the 1st recursive `mergeSort` while satisfying `@pre`.
3. Assuming the 1st recursive call returns with `@post` satisfied, the 2nd recursive `mergeSort` is entered while satisfying `@pre`.
4. Assuming the 2nd recursive call returns with `@post` satisfied, once the original function has terminated, it satisfies `@post`.
(That is, of course assuming this function terminates.)

**Comment by KihongHeo** [comment]
Exactly.

**Comment by pingpingy1** [comment]
That's great!
Thank you very much.
